src/logs.rs

```diff
@@ -0,0 +1,149 @@
+    }
+}
+
+impl io::Write for LogStream {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.append(buf)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl LogStream {
+    #[must_use]
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            capacity,
+            ..Default::default()
+        }
+    }
+
+    /// Append a buffer to the log stream and truncates when hitting the capacity.
+    /// We return the input buffer size regardless of whether we truncated or not to avoid a panic.
+    /// # Arguments
+    /// * `buf` - the buffer to append
+    /// # Returns
+    /// * `Ok(usize)` - the number of bytes in the buffer that was passed in
+    /// * `Err(io::Error)` - if the buffer is empty
+    /// # Errors
+    /// * `io::Error` - if the buffer is empty
+    pub fn append(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if self.current_bytesize > self.capacity {
+            return Ok(buf.len());
+        }
+
+        if buf.is_empty() {
+            return Ok(0);
+        }
+
+        let log = String::from_utf8_lossy(buf);
+        let (truncated, log) =
+            truncate_to_char_boundary(&log, self.capacity - self.current_bytesize);
+        let mut log = log.to_string();
+        if truncated {
+            log.push_str("...[TRUNCATED]");
+        }
+
+        let size = log.len();
+
+        self.current_bytesize += size;
+        self.logs.push(log);
+
+        Ok(buf.len())
+    }
+
+    #[must_use]
+    pub fn last(&self) -> Option<&String> {
+        self.logs.last()
+    }
+
+    #[must_use]
+    pub fn last_message(&self) -> Option<&str> {
+        self.logs.last().map(String::as_str)
+    }
+}
+
+// truncate `&str` to length at most equal to `max`
+// return `true` if it were truncated, and the new str.
+fn truncate_to_char_boundary(s: &str, mut max: usize) -> (bool, &str) {
+    if max >= s.len() {
+        (false, s)
+    } else {
+        while !s.is_char_boundary(max) {
+            max -= 1;
+        }
+        (true, &s[..max])
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_bounded_log() {
+        let mut bounded_log = LogStream::with_capacity(15);
+        let log = b"hello world";
+        bounded_log.append(log).unwrap();
+        assert_eq!(Some("hello world"), bounded_log.last_message());
+    }
+
+    #[test]
+    fn test_bounded_log_when_truncated() {
+        let mut bounded_log = LogStream::with_capacity(10);
+        let log = b"hello world";
+        bounded_log.append(log).unwrap();
+        assert_eq!(Some("hello worl...[TRUNCATED]"), bounded_log.last_message());
+    }
+
+    #[test]
+    fn test_bounded_log_when_truncated_nearest_valid_utf8() {
+        let mut bounded_log = LogStream::with_capacity(15);
+        bounded_log.append("✌️✌️✌️".as_bytes()).unwrap(); // ✌️ is 6 bytes, ✌ is 3;
+        assert_eq!(Some("✌️✌️✌...[TRUNCATED]"), bounded_log.last_message());
```

Comment:
Why is it saying truncated if the output before `...[TRUNCATED]` is the same as what was logged (there wasn't any truncation)?

Reply:
the 3 peace signs are 6 bytes each, and look like it crops the last one to the 3 bytes variant